File: src/ir/basic_block.cpp
#include "toyc/ir/basic_block.h"

BasicBlock::BasicBlock(Label* label)
    : label_(label) {}

void BasicBlock::add_phi(InstrPtr phi) {
    phis_.push_back(std::move(phi));
}

void BasicBlock::clear_phis() {
    phis_.clear();
}

void BasicBlock::insert_instruction(size_t index, InstrPtr instr) {
    if (is_terminated()) return;
    if (index > instrs_.size()) index = instrs_.size();
    instrs_.insert(instrs_.begin() + index, std::move(instr));
}

void BasicBlock::add_instruction(InstrPtr instr) {
    if (is_terminated()) {
        // Mimics Java: "cannot append instruction after terminator"
        // For now, silently ignore; caller should check.
        return;
    }
    instrs_.push_back(std::move(instr));
}

void BasicBlock::replace_body(std::vector<InstrPtr> new_instrs) {
    instrs_ = std::move(new_instrs);
}

void BasicBlock::set_terminator(InstrPtr term) {
    // Java validates isTerminator() here; we trust the caller for now.
    terminator_ = std::move(term);
}

std::vector<Instr*> BasicBlock::all_instrs() const {
    std::vector<Instr*> result;
    result.reserve(phis_.size() + instrs_.size() + (terminator_ ? 1 : 0));
    for (const auto& p : phis_) {
        result.push_back(p.get());
    }
    for (const auto& i : instrs_) {
        result.push_back(i.get());
    }
    if (terminator_) {
        result.push_back(terminator_.get());
    }
    return result;
}

File: toy-c-compiler-master/src/main/java/toyc/ir/block/BasicBlock.java
package toyc.ir.block;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import toyc.ir.IRVisitor;
import toyc.ir.inst.Instruction;
import toyc.ir.inst.Phi;
import toyc.ir.value.Label;

public final class BasicBlock {
    private final Label label;
    private final List<Phi> phis = new ArrayList<>();
    private final List<Instruction> instructions = new ArrayList<>();
    private Instruction terminator;

    public BasicBlock(Label label) {
        this.label = label;
    }

    public Label label() {
        return label;
    }

    public List<Phi> phis() {
        return Collections.unmodifiableList(phis);
    }

    public List<Instruction> instructions() {
        return Collections.unmodifiableList(instructions);
    }

    public List<Instruction> allInstructions() {
        List<Instruction> all = new ArrayList<>(phis.size() + instructions.size() + (terminator == null ? 0 : 1));
        all.addAll(phis);
        all.addAll(instructions);
        if (terminator != null) {
            all.add(terminator);
        }
        return Collections.unmodifiableList(all);
    }

    public void addPhi(Phi phi) {
        phis.add(phi);
    }

    public void clearPhis() {
        phis.clear();
    }

    public void addInstruction(Instruction instruction) {
        if (isTerminated()) {
            throw new IllegalStateException("cannot append instruction after terminator in " + label);
        }
        if (instruction instanceof Phi phi) {
            addPhi(phi);
        } else if (instruction.isTerminator()) {
            terminator = instruction;
        } else {
            instructions.add(instruction);
        }
    }

    public void replaceInstructions(List<Instruction> newInstructions) {
        phis.clear();
        instructions.clear();
        terminator = null;
        for (Instruction instruction : newInstructions) {
            addInstruction(instruction);
        }
    }

    public void replaceBody(List<Instruction> newInstructions) {
        for (Instruction instruction : newInstructions) {
            if (instruction instanceof Phi || instruction.isTerminator()) {
                throw new IllegalArgumentException("block body cannot contain phi or terminator");
            }
        }
        instructions.clear();
        instructions.addAll(newInstructions);
    }

    public void setTerminator(Instruction terminator) {
        if (terminator != null && !terminator.isTerminator()) {
            throw new IllegalArgumentException("terminator instruction expected");
        }
        this.terminator = terminator;
    }

    public boolean isTerminated() {
        return terminator != null;
    }

    public Instruction terminator() {
        return terminator;
    }

    public boolean hasTerminatorType(Class<? extends Instruction> type) {
        return terminator != null && type.isInstance(terminator);
    }

    public <R, C> R accept(IRVisitor<R, C> visitor, C context) {
        return visitor.visitBasicBlock(this, context);
    }
}
